README.md

```diff
@@ -0,0 +1,492 @@
+!["cached flow"](docs/CachedValue.svg)
+
+##### Invalid CacheValue
+when you call `run` or `subscribe` on a `query` with a combination of `inputs` that was never used before (or whose last use ended up with a `Failure`), avenger will try to run the `Fetch` function resulting in a more complex flow:
+!["cached flow"](docs/UncachedOrErrorValue.svg)
+
+There are two ways to get a query result:
+
+```ts
+type Error = "500" | "404";
+type User = { userName: String };
+
+declare function getUser(userId: number): TaskEither<Error, User>
+
+const userQuery: CachedQuery<number, Error, User> = query(
+  getUser,
+  refetch
+)(refetch);
+
+declare function dispatchError(e: Error): void;
+declare function setCurrentUser(e: User): void;
+
+// feeding your query to `observe` will give you an observable on the query
+// N.B. until now no fetch is yet attempted, avenger will wait until the first subscription is issued
+const observable: Observable<QueryResult<Error, User>> = observe(userQuery);
+
+// this will trigger the fetch function
+observable.subscribe(dispatchError, setCurrentUser);
+
+// alternatively you can call `run` on your query and it will return a TaskEither<Error, User>
+// you can then use it imperatively
+const task: TaskEither<Error, User> = userQuery.run(1);
+const result: Either<Error, User> = await task.run();
+```
+
+although the `run` method is available to check a query result imperatively, it is highly suggested the use of the `observe` utility in order to be notified in real time of when data changes.
+
+Either way, whenever you ask for a query result you will end up with an object with the [**`QueryResult`**](#QueryResult) signature that conveniently lets you fold to decide the best way to handle the result. The `fold` method takes three functions as parameters: the first is used to handle a `Loading` result; the second is used in case a `Failure` occurs; the last one handles `Success` values.
+
+
+## composing queries
+You can build bigger queries from smaller ones in two ways:
+
+- by composing them with [**`compose`**](#compose): when you need your queries to be sequentially run with the results of one feeding the other, you can use `compose`.
+- by grouping them with [**`product`**](#product): when you don't need to run the queries sequentially but would like to conveniently group them and treat them as if they were one you can use `product`*.
+
+*Internally `product` uses the `Applicative` nature of `QueryResults` to group them using the following hierarchical logic:
+  1. If any of the queries returned a `Failure` then the whole composition is a `Failure`.
+  2. If any of the queries is `Loading` then the whole composition is `Loading`.
+  3. If all the queries ended with a `Success` then the composition is a `Success` with a record of results that mirrors the key/value result of the single queries as value.
+
+Here are a couple of simple examples on how to use them:
+
+```ts
+import { compose } from 'avenger/lib/Query';
```

Comment:
I would say something like "// each value defined is explicitly annotated for clarity, although the annotations are not strictly required"